xchange-lgo/src/main/java/org/knowm/xchange/lgo/LgoAdapters.java

```diff
@@ -0,0 +1,179 @@
+              Currency.USD));
+    }
+    return metaData;
+  }
+
+  private static CurrencyPair toPair(LgoProduct product) {
+    return new CurrencyPair(
+        Currency.getInstance(product.getBase().getId()),
+        Currency.getInstance(product.getQuote().getId()));
+  }
+
+  public static Stream<LgoKey> adaptKeysIndex(String indexFile) {
+    String[] all = indexFile.split("(true|false)");
+    Spliterator<String> spliterator =
+        Spliterators.spliterator(all, Spliterator.IMMUTABLE | Spliterator.NONNULL);
+    return StreamSupport.stream(spliterator, true).map(LgoAdapters::parseSummary);
+  }
+
+  private static LgoKey parseSummary(String rawSummary) {
+    String[] summary = rawSummary.split(" ");
+    return new LgoKey(summary[0], Instant.parse(summary[1]), Instant.parse(summary[2]));
+  }
+
+  public static LgoPlaceOrder adaptLimitOrder(LimitOrder limitOrder) {
+    String product = adaptCurrencyPair(limitOrder.getCurrencyPair());
+    String side = adaptOrderType(limitOrder.getType());
+    return new LgoPlaceLimitOrder(
+        0,
+        side,
+        product,
+        limitOrder.getOriginalAmount(),
+        limitOrder.getLimitPrice(),
+        limitOrder.getTimestamp().toInstant());
+  }
+
+  public static LgoPlaceOrder adaptMarketOrder(MarketOrder marketOrder) {
+    String product = adaptCurrencyPair(marketOrder.getCurrencyPair());
+    String side = adaptOrderType(marketOrder.getType());
+    return new LgoPlaceMarketOrder(
+        0, side, product, marketOrder.getOriginalAmount(), marketOrder.getTimestamp().toInstant());
+  }
+
+  public static String adaptOrderType(OrderType type) {
+    return type == OrderType.BID ? "B" : "S";
+  }
+
+  public static String adaptCurrencyPair(CurrencyPair currencyPair) {
+    return String.format(
+        "%s-%s", currencyPair.base.getCurrencyCode(), currencyPair.counter.getCurrencyCode());
+  }
+
+  private static UserTrade adaptUserTrade(LgoUserTrade lgoUserTrade) {
+    OrderType type = adaptUserTradeType(lgoUserTrade);
+    CurrencyPair currencyPair = adaptProductId(lgoUserTrade.getProductId());
+    Date creationDate = adaptDate(lgoUserTrade.getCreationDate());
+    BigDecimal originalAmount = new BigDecimal(lgoUserTrade.getQuantity());
+    BigDecimal price = new BigDecimal(lgoUserTrade.getPrice());
+    BigDecimal feeAmount = new BigDecimal(lgoUserTrade.getFees());
+    return new UserTrade(
+        type,
+        originalAmount,
+        currencyPair,
+        price,
+        creationDate,
+        lgoUserTrade.getId(),
+        lgoUserTrade.getOrderId(),
+        feeAmount,
+        currencyPair.counter);
+  }
+
+  public static CurrencyPair adaptProductId(String productId) {
+    String[] pair = productId.split("-");
+    return new CurrencyPair(pair[0], pair[1]);
+  }
+
+  public static Date adaptDate(String date) {
+    try {
+      return dateFormat.parse(date);
+    } catch (ParseException e) {
+      LOG.error("Error while parsing trade date", e);
```

Comment:
throwing an exception here would be better then getting a NPE in your application code later

Reply:
👍